src/Microsoft.AspNet.Razor/CodeGenerators/CSharpTagHelperCodeRenderer.cs

```diff
@@ -251,69 +226,134 @@ protected virtual string GenerateUniqueId()
+                    descriptor.Attributes.Any(attributeDescriptor => attributeDescriptor.IsNameMatch(attribute.Key)));
+
+                // Bound attributes will have associated descriptors.
+                if (associatedDescriptors.Any())
                 {
-                    // We only need to create attribute values once per HTML element (not once per tag helper).
-                    // We're saving the value accessor so we can retrieve it later if there are more tag
-                    // helpers that need the value.
-                    htmlAttributeValues.Add(attributeName, valueAccessor);
-
-                    // Bufferable attributes are attributes that can have Razor code inside of them. Such
-                    // attributes have string values and may be calculated using a temporary TextWriter or other
-                    // buffer.
-                    var bufferableAttribute = attributeDescriptor.IsStringProperty;
-
-                    RenderNewAttributeValueAssignment(
-                        attributeDescriptor,
-                        bufferableAttribute,
-                        attributeValueChunk,
-                        valueAccessor);
-
-                    // Execution contexts are a runtime feature.
-                    if (_designTimeMode)
+                    if (!renderedBoundAttributeNames.Add(attribute.Key))
                     {
+                        // We already rendered an attribute with the same name that was bound. First wins, skip.
                         continue;
                     }
 
-                    // We need to inform the context of the attribute value.
-                    _writer
-                        .WriteStartInstanceMethodInvocation(
-                            ExecutionContextVariableName,
-                            _tagHelperContext.ExecutionContextAddTagHelperAttributeMethodName)
-                        .WriteStringLiteral(attributeName)
-                        .WriteParameterSeparator()
-                        .Write(valueAccessor)
-                        .WriteEndMethodInvocation();
+                    // We capture the value location of each bound attribute to ensure when multiple TagHelpers request
+                    // the same attribute we don't execute the attributes value more than once.
+                    string valueLocation = null;
+
+                    foreach (var associatedDescriptor in associatedDescriptors)
+                    {
+                        // There should only be a single TagHelperAttributeDescriptor associated with the attribute.
+                        // TagHelpers cannot have multiple attribute descriptors pointing to the same property unless
+                        // they're indexed.
+                        var associatedAttributeDescriptor = associatedDescriptor.Attributes.First(
+                            attributeDescriptor => attributeDescriptor.IsNameMatch(attribute.Key));
```

Comment:
Do not "fix" the _separate_ bug about multiple matching attribute descriptors in this way. Leads to unfathomable errors from Linq because `TagHelperBlockRewriter` is not guarding against this situation. Stick with `FirstOrDefault()` and leave #199 alone. (#237 also touches this case but that enhancement is for later.)